File: Android/mobile/src/main/java/com/windkracht8/rugbyrefereewatch/DeviceConnect.java
/*
 * Copyright 2020-2025 Bart Vullings <dev@windkracht8.com>
 * This file is part of RugbyRefereeWatch
 * RugbyRefereeWatch is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * RugbyRefereeWatch is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.windkracht8.rugbyrefereewatch;

import android.app.Activity;
import android.graphics.drawable.AnimatedVectorDrawable;
import android.os.Bundle;
import android.util.Log;
import android.widget.ImageView;
import android.widget.TextView;

import org.json.JSONObject;

public class DeviceConnect extends Activity implements Comms.Interface{
    @Override public void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        setContentView(R.layout.device_connect);
        findViewById(android.R.id.content).setOnApplyWindowInsetsListener(Main.onApplyWindowInsetsListener);
        String name = getIntent().getStringExtra("name");
        if(name == null) ((TextView)findViewById(R.id.device_connect_name)).setText(R.string.connecting_to);
        ((TextView)findViewById(R.id.device_connect_name)).setText(getString(R.string.connecting_to, name));

        ImageView icon = findViewById(R.id.device_connect_icon);
        ((AnimatedVectorDrawable) icon.getBackground()).start();

        if(Main.comms == null || Main.comms.status != Comms.Status.CONNECTING) finishAndRemoveTask();
        try{
            Main.comms.addListener(this);
        }catch(Exception e){
            Log.e(Main.LOG_TAG, "DeviceConnect.onCreate Failed to add as a listener: " + e.getMessage());
        }
    }
    @Override public void onDestroy(){
        super.onDestroy();
        if(Main.comms != null) Main.comms.removeListener(this);
    }
    @Override public void onCommsStartDone(){finishAndRemoveTask();}
    @Override public void onCommsConnecting(String x){}
    @Override public void onCommsConnectFailed(){finishAndRemoveTask();}
    @Override public void onCommsConnected(String x){finishAndRemoveTask();}
    @Override public void onCommsSending(){}
    @Override public void onCommsSendingFinished(){}
    @Override public void onCommsDisconnected(){finishAndRemoveTask();}
    @Override public void onCommsResponse(Comms.Request.Type t, JSONObject r){}
    @Override public void onCommsError(int e){}
}

File: Android/mobile/src/main/java/DeviceConnect.kt
/*
 * Copyright 2020-2026 Bart Vullings <dev@windkracht8.com>
 * This file is part of RugbyRefereeWatch
 * RugbyRefereeWatch is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 * RugbyRefereeWatch is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.windkracht8.rugbyrefereewatch

import android.content.res.Configuration
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.animation.graphics.res.animatedVectorResource
import androidx.compose.animation.graphics.res.rememberAnimatedVectorPainter
import androidx.compose.animation.graphics.vector.AnimatedImageVector
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.safeDrawingPadding
import androidx.compose.material3.MaterialTheme.colorScheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.sp
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.launch

class DeviceConnect : ComponentActivity() {
	public override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		if (Comms.status.value != Comms.Status.CONNECTING) finishAndRemoveTask()
		lifecycleScope.launch { Comms.status.collect {
			if (it != Comms.Status.CONNECTING) finishAndRemoveTask()
		} }
		setContent { W8Theme (window, resources) { Surface { DeviceConnectScreen() } } }
	}
}

@Composable
fun DeviceConnectScreen() {
	val iconAnimation = AnimatedImageVector.animatedVectorResource(R.drawable.watch_connecting)
	var iconAnimationAtEnd by remember { mutableStateOf(false) }
	Column(Modifier.fillMaxSize().safeDrawingPadding()) {
		Text(
			modifier = Modifier.fillMaxWidth(),
			text = stringResource(R.string.connecting_to, Comms.deviceName),
			textAlign = TextAlign.Center,
			color = colorScheme.primary,
			fontSize = 20.sp,
			fontWeight = FontWeight.Bold
		)
		Text(
			modifier = Modifier.fillMaxWidth(),
			text = stringResource(R.string.device_connect_instruct),
			textAlign = TextAlign.Center,
			fontSize = 14.sp,
			fontWeight = FontWeight.Bold
		)
		Image(
			modifier = Modifier.fillMaxSize(),
			painter = rememberAnimatedVectorPainter(iconAnimation, iconAnimationAtEnd),
			contentDescription = "watch icon"
		)
		iconAnimationAtEnd = true
	}
}

@Preview(uiMode = Configuration.UI_MODE_NIGHT_YES, device = Devices.PIXEL_8)
@Composable
fun PreviewDeviceConnect() {
	Comms
	W8Theme (null, null) { Surface { DeviceConnectScreen() } }
}
@Preview(device = Devices.PIXEL_8)
@Composable
fun PreviewDeviceConnectDay() { PreviewDeviceConnect() }
